Fix the type of random seed from i64 to i32.

whchung opened the pull request:
Match what's implied in mlir-miopen-driver CPU logic.

This fixes an weird bug that CPU convolution may have different results in different runs, under a fixed random seed value.

whchung committed:
Fix the type of random seed from i64 to i32.

Match what's implied in mlir-miopen-driver CPU logic.

whchung commented:
https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/miopen-dialect/mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp#L1516

Is where `mlir-miopen-driver` assumes the `seed` value be an `i32` instead of `i64`.

jerryyin commented on mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp:
```diff
@@ -499,7 +499,7 @@ mcpuMemset5DFloatRandInt(float *allocated, float *aligned, int64_t offset,
                          int64_t size3, int64_t size4, int64_t stride0,
                          int64_t stride1, int64_t stride2, int64_t stride3,
-                         int64_t stride4, short min, short max, int64_t seed) {
+                         int64_t stride4, short min, short max, int seed) {
```
Maybe go with `int32_t` to be conservative?

jungpark-mlir replied:
https://en.cppreference.com/w/cpp/numeric/random/srand
Looks srand expects 'unsigned' as seed type, I guess better not specify the size then.

jerryyin replied:
@jungpark-mlir Problem is that the populated test IR will generate a i32. I'm indicating to match based on that. But your observation is also valid. Guess this chunk of code was just poorly written. The populated code as well as the function signature should both be unsigned.

At this case, I'm less worried about the stl function's signature because it will do a implicit type conversion. However the runtime function matters more because we have no such cast in the populated IR.

jungpark-mlir replied:
I see, thanks!

whchung replied:
@jerryyin / @jungpark-mlir I’ll amend the PR so:

1) rand in mlir-miopen-driver is always a positive integer or 0.
2) in case rand is 0, use srand(time()), otherwise use the specified number as the random seed.
3) change relevant documentation amd test cases.
4) use uint32_t here.

how do you think?

jerryyin replied:
@whchung Thanks, that sounds much better. We wouldn't be accidently pass a negative to srand after your change.

whchung committed:
Revise -rand flag in mlir-miopen-driver.

- -rand in mlir-miopen-driver is always a positive integer or 0.
- In case -rand is 0, use srand(time()), otherwise use the specified number as the random seed.
- If not specified, use all 1.
- Change relevant documentation.
- Use uint32_t in rocm-runtime-wrappers.

whchung commented:
@jerryyin please help review this PR again.

jerryyin commented on mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp:
```diff
@@ -499,8 +499,8 @@ mcpuMemset5DFloatRandInt(float *allocated, float *aligned, int64_t offset,
                          int64_t stride1, int64_t stride2, int64_t stride3,
-                         int64_t stride4, short min, short max, int seed) {
-  if (seed < 0)
+                         int64_t stride4, short min, short max, uint32_t seed) {
```
I think to do it 100% right we need to populate the `ui32` type instead of `i32` type in the `mlir-miopen-driver.cpp` end. But `i32` does not concern about signed-ness so is correct too.

whchung commented:
jenkins: retest this please.
